sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,305 @@
+//! The networkd_status module contains definitions and functions for tracking network status from networkd
+//!
+//! The primary purpose of this module is to provide structures to read in a `networkctl status` command output
+//! and make specific fields available from this output. These structs can then be used to read DNS, IP Addressing
+//! and any other networking status data for use in configuration files needed for other networking tools.
+use crate::interface_id::InterfaceName;
+use crate::NETWORKCTL;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer};
+use snafu::{ensure, ResultExt};
+use std::cmp::Ordering;
+use std::convert::TryInto;
+use std::net::IpAddr;
+use std::process::Command;
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+pub(crate) struct NetworkDInterfaceStatus {
+    pub(crate) name: InterfaceName,
+    #[serde(rename = "DNS")]
+    pub(crate) dns: Option<Vec<NetworkDDnsConfig>>,
+    pub(crate) search_domains: Option<Vec<SearchDomain>>,
+    #[serde(rename = "Addresses", deserialize_with = "from_networkctl_addresses")]
+    pub(crate) addresses: Vec<IpAddr>,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct NetworkDDnsConfig {
+    #[serde(rename = "Address", deserialize_with = "ipaddr_from_vec_de")]
+    pub(crate) address: IpAddr,
+}
+
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct SearchDomain {
+    #[serde(rename = "Domain")]
+    pub(crate) domain: String,
+}
+
+// get an IpAddr from a Vec<u8> (could be 4 or 16 length)
+fn ipaddr_from_vec(address_vec: Vec<u8>) -> Result<IpAddr> {
+    match address_vec.len() {
+        4 => {
+            // We checked that its exactly 4 u8 long, forcing std::result::Result explicity to raise a nice error
+            let octet_result: std::result::Result<[u8; 4], _> = address_vec.try_into();
+            match octet_result {
+                Ok(octets) => Ok(IpAddr::from(octets)),
+                Err(_) => error::BadIpAddressVecSnafu.fail(),
+            }
+        }
+        16 => {
+            // We checked that its exactly 16 u8 long, forcing std::result::Result explicity to raise a nice error
+            let octet_result: std::result::Result<[u8; 16], _> = address_vec.try_into();
+            match octet_result {
+                Ok(octets) => Ok(IpAddr::from(octets)),
+                Err(_) => error::BadIpAddressVecSnafu.fail(),
+            }
+        }
```

Comment:
You can use `expect()` or `unwrap()` here given that you've tested the length:
```suggestion
        4 => Ok(IpAddr::from(TryInto::<[u8; 4]>::try_into(address_vec).expect("4 bytes"))),
        16 => Ok(IpAddr::from(TryInto::<[u8; 16]>::try_into(address_vec).expect("16 bytes"))),
```

Reply:
We went back and forth on if `.expect()` was acceptable (and a previous revision had that logic instead). I like the fact we are now avoiding a `panic` but the code is much less readable than `expect()`. I'm ok with this but would like to have @zmrow weigh in just to confirm we are all on the same page.